### PR TITLE
Reduce boilerplate with timelines

### PR DESCRIPTION
**Pull request by marrus-sh**
Right now the Home, Community, Public, and Hashtag timelines all use basically the same code with a lot of boilerplate. This creates a new `<Timeline>` component to make maintenance of these easier.

### CODE REVIEW

**Comment by beatrix-bitrot** [comment]
this is one of those ones that i like a lot... but want in upstream rather than glitchsoc because i fear the potential for nasty business when upstream changes these things

do you think it would be accepted?

**Comment by nightpool** [comment]
I don't touch that code much—I haven't actually done anything significant with react yet. I would worry about over-DRYing the code though. This refactor seems like a lot of code churn for very little benefit. I would maybe suggest figure out a object to put that subscription code in, as a good first step towards better abstraction?

If you do want to make this refactor though, why not use inheritance? seems like a perfect situation for it.

**Comment by marrus-sh** [comment]
The benefit is that it allows all timelines to share the same code? For example, right now (afaict) the little blue indicator light for new toots is broken on hashtag timelines, because nobody updated the indicator light code on the hashtag timeline component when the timeline-id's of hashtag timelines was changed from just `tag` to `hashtag:${id}`. If hashtag timelines were implemented the way they are here, that wouldn't be a problem, because you would only have to change the variable once (in the prop).

The same goes for stream id's, column names, etc. Instead of declaring these dozens of times in the same file you just declare it once when you use the component. And it makes it far easier to add additional timeline types in the future.

> why not use inheritance?

“React has a powerful composition model, and we recommend using composition instead of inheritance to reuse code between components.” :P https://facebook.github.io/react/docs/composition-vs-inheritance.html

File: app/javascript/mastodon/features/hashtag_timeline/index.js [hunk 1]
```diff
@@ -1,138 +1,53 @@
...
-          scrollKey={`hashtag_timeline-${columnId}`}
-          timelineId={`hashtag:${id}`}
-          loadMore={this.handleLoadMore}
-          emptyMessage={<FormattedMessage id='empty_column.hashtag' defaultMessage='There is nothing in this hashtag yet.' />}
-        />
-      </Column>
+      <Timeline
+        expand={this.expand}
+        refresh={this.refresh}
+        streamId={`hashtag&tag=${id}`}
```
Comment (nightpool) [inline app/javascript/mastodon/features/hashtag_timeline/index.js]: this isn't an id, it's a path fragment

File: app/javascript/mastodon/features/community_timeline/index.js [hunk 2]
```diff
@@ -1,144 +1,44 @@
...
-        />
-      </Column>
+      <Timeline
+        expand={expandCommunityTimeline}
+        refresh={refreshCommunityTimeline}
+        streamId='public:local'
+        columnName='COMMUNITY'
+        columnId={columnId}
+        mulitColumn={multiColumn}
+        emptyMessage={<FormattedMessage id='empty_column.community' defaultMessage='The local timeline is empty. Write something publicly to get the ball rolling!' />}
```
Comment (nightpool) [inline app/javascript/mastodon/features/community_timeline/index.js]: why is this a formattedmessage instead of an intl.formatMessage as below?

**Comment by nightpool** [comment]
well, what I can tell you is that it feels like over-DRY code that will be hard to specialize down the line. Maybe that's not valid, but the amount of random properties on Timeline that don't seem to have a coherent semantic meaning makes me very cautious. (why is community in there 3 times, for example?)

File: app/javascript/mastodon/features/community_timeline/index.js [hunk 3]
```diff
@@ -1,144 +1,44 @@
...
+        icon='users'
+        title={intl.formatMessage(messages.title)}
+        settings={<ColumnSettingsContainer />}
+        scrollName='community_timeline'
+        timelineId='community'
```
Comment (nightpool) [inline app/javascript/mastodon/features/community_timeline/index.js]: columnName, scrollName, and timelineId are all basically redundant—they should either be the same or programmatically generated from a single id.

File: app/javascript/mastodon/features/community_timeline/index.js [hunk 4]
```diff
@@ -1,144 +1,44 @@
...
-          scrollKey={`community_timeline-${columnId}`}
-          timelineId='community'
-          loadMore={this.handleLoadMore}
-          emptyMessage={<FormattedMessage id='empty_column.community' defaultMessage='The local timeline is empty. Write something publicly to get the ball rolling!' />}
-        />
-      </Column>
+      <Timeline
+        expand={expandCommunityTimeline}
+        refresh={refreshCommunityTimeline}
+        streamId='public:local'
```
Comment (nightpool) [inline app/javascript/mastodon/features/community_timeline/index.js]: this should be on the action side of things, in the same way refreshCommunityTimeline is set up.

File: app/javascript/mastodon/features/hashtag_timeline/index.js (hunk 1 above)
Comment (nightpool) [inline app/javascript/mastodon/features/hashtag_timeline/index.js]: also see above

File: app/javascript/mastodon/features/timeline/index.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    emptyMessage: PropTypes.oneOfType([
+      PropTypes.element,
+      PropTypes.string,
+    ]),
+    icon: PropTypes.string.isRequired,
+    title: PropTypes.string.isRequired,
+    settings: PropTypes.element,
+    scrollName: PropTypes.string.isRequired,
+    timelineId: PropTypes.string.isRequired,
+  };
```
Comment (nightpool) [inline app/javascript/mastodon/features/timeline/index.js]: at 19 lines, this is the second longest propTypes declaration in the entire code base. that's what makes me worried this is a bad refactor—if we have to do this much specialization now, what happens when we want to make things more complicated? Huge, clear cut violation of the open/close principle.

File: app/javascript/mastodon/features/timeline/index.js [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    this.props.dispatch(this.props.expand());
+  }
+
+  _subscribe (dispatch, streamId, timelineId) {
+    const { streamingAPIBaseURL, accessToken } = this.props;
+
+    if (!streamId || !timelineId) return;
+
+    this.subscription = createStream(streamingAPIBaseURL, accessToken, streamId, {
+
```
Comment (nightpool) [inline app/javascript/mastodon/features/timeline/index.js]: no newline

File: app/javascript/mastodon/features/timeline/index.js [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        switch(data.event) {
+        case 'update':
+          dispatch(updateTimeline(timelineId, JSON.parse(data.payload)));
+          break;
+        case 'delete':
+          dispatch(deleteFromTimelines(data.payload));
+          break;
+        }
+      },
+
```
Comment (nightpool) [inline app/javascript/mastodon/features/timeline/index.js]: no newline

**Comment by beatrix-bitrot** [comment]
I wonder, now that we have separated out glitch things, would it make sense to do the refactor there and leave the upstream side unchanged?

**Pull request closed without merging**